crates/astria-proto/src/native/sequencer/v1alpha1.rs

```diff
@@ -1251,6 +1266,1158 @@ impl Display for IncorrectAddressLength {
+        }
+
+        Ok(Self {
+            block_hash,
+            header,
+            rollup_ids,
+            rollup_transactions_root,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+        })
+    }
+
+    /// Converts into the raw decoded protobuf representation of this type.
+    pub fn into_raw(self) -> raw::CelestiaSequencerBlob {
+        let Self {
+            header,
+            rollup_ids,
+            rollup_transactions_root,
+            rollup_transactions_proof,
+            rollup_ids_proof,
+            ..
+        } = self;
+        raw::CelestiaSequencerBlob {
+            header: Some(header.into()),
+            rollup_ids: rollup_ids.into_iter().map(RollupId::to_vec).collect(),
+            rollup_transactions_root: rollup_transactions_root.to_vec(),
+            rollup_transactions_proof: Some(rollup_transactions_proof.into_raw()),
+            rollup_ids_proof: Some(rollup_ids_proof.into_raw()),
+        }
+    }
+
+    /// Converts from the raw decoded protobuf representation of this type.
+    ///
+    /// # Errors
+    /// TODO(https://github.com/astriaorg/astria/issues/612)
+    pub fn try_from_raw(
+        raw: raw::CelestiaSequencerBlob,
+    ) -> Result<Self, CelestiaSequencerBlobError> {
+        UncheckedCelestiaSequencerBlob::try_from_raw(raw)
+            .and_then(UncheckedCelestiaSequencerBlob::try_into_celestia_sequencer_blob)
+    }
+}
+
+fn are_rollup_ids_included<'a, TRollupIds: 'a>(
+    ids: TRollupIds,
+    proof: &merkle::Proof,
+    data_hash: [u8; 32],
+) -> bool
+where
+    TRollupIds: IntoIterator<Item = RollupId>,
+{
+    let tree = merkle::Tree::from_leaves(ids);
+    let hash_of_root = Sha256::digest(tree.root());
+    proof.verify(&hash_of_root, data_hash)
+}
+
+fn are_rollup_txs_included(
+    rollup_txs: &IndexMap<RollupId, Vec<Vec<u8>>>,
+    rollup_proof: &merkle::Proof,
+    data_hash: [u8; 32],
+) -> bool {
+    let rollup_tree = derive_merkle_tree_from_rollup_txs(rollup_txs);
+    let hash_of_rollup_root = Sha256::digest(rollup_tree.root());
+    rollup_proof.verify(&hash_of_rollup_root, data_hash)
+}
+
+/// Derive a [`merkle::Tree`] from an iterable.
+///
+/// It is the responsbility if the caller to ensure that the iterable is
+/// deterministic. Prefer types like `Vec`, `BTreeMap` or `IndexMap` over
+/// `HashMap`.
+pub fn derive_merkle_tree_from_rollup_txs<'a, T: 'a>(rollup_ids_to_txs: T) -> merkle::Tree
+where
+    T: IntoIterator<Item = (&'a RollupId, &'a Vec<Vec<u8>>)>,
+{
+    let mut tree = merkle::Tree::new();
+    for (rollup_id, txs) in rollup_ids_to_txs {
+        let root = merkle::Tree::from_leaves(txs).root();
+        tree.build_leaf().write(rollup_id.as_ref()).write(&root);
+    }
+    tree
+}
+
+// TODO: This can all be done in-place once https://github.com/rust-lang/rust/issues/80552 is stabilized.
+pub fn group_sequence_actions_in_signed_transaction_transactions_by_rollup_id(
```

Comment:
This was moved/adapted from astria-sequencer. Because other parts of the code are interacting with it (mainly through tests) it made sense to me to put it here, also because it's closely relatted to `derive_merkle_from_rollup_txes`